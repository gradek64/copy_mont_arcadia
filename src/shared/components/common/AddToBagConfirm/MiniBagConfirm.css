/** @define MiniBagConfirm */

.MiniBagConfirm {
  position: absolute;
  top: 100%;
  z-index: 1;
  pointer-events: none;
  transition: all 1s ease-in;

  &.is-visible {
    opacity: 1;
    pointer-events: auto;
  }

  &-content {
    padding: 20px 10px 10px;
  }

  &-closeIcon {
    &:before {
      content: 'Ã—'
    }
  }

  &-products {
    margin: 0;
    padding: 0;
  }

  &-goToCheckout {
    width: 100%;
  }

  &-viewBag {
    width: 100%;
    margin-top: 5px;
  }
}
