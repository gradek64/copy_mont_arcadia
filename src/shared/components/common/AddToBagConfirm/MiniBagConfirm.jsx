import classnames from 'classnames'
import PropTypes from 'prop-types'
import { propEq, sort } from 'ramda'
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { browserHistory } from 'react-router'

// actions
import {
  openMiniBag,
  showMiniBagConfirm,
} from '../../../actions/common/shoppingBagActions'

// components
import OrderProduct from '../OrderProducts/OrderProduct'
import ToolTip from '../ToolTip/ToolTip'
import Button from '../Button/Button'

class MiniBagConfirm extends Component {
  static propTypes = {
    isShown: PropTypes.bool,
    quantity: PropTypes.number,
    products: PropTypes.array,
    redirectTo: PropTypes.func,
    setTimeout: PropTypes.func,
    clearTimeout: PropTypes.func,
    // actions
    showMiniBagConfirm: PropTypes.func.isRequired,
    openMiniBag: PropTypes.func.isRequired,
  }

  static defaultProps = {
    isShown: false,
    quantity: 0,
    products: [],
    redirectTo:
      typeof browserHistory !== 'undefined' && 'push' in browserHistory
        ? browserHistory.push
        : () => {},
    setTimeout:
      typeof window !== 'undefined' && 'setTimeout' in window
        ? window.setTimeout.bind(window)
        : () => {},
    clearTimeout:
      typeof window !== 'undefined' && 'clearTimeout' in window
        ? window.clearTimeout.bind(window)
        : () => {},
  }

  static contextTypes = {
    l: PropTypes.func,
  }

  timeoutID = undefined

  componentDidUpdate(prevProps) {
    const { isShown, setTimeout, showMiniBagConfirm } = this.props
    if (!prevProps.isShown && isShown) {
      this.timeoutID = setTimeout(() => {
        showMiniBagConfirm(false)
        this.timeoutID = undefined
      }, 6000)
    }
  }

  componentWillUnmount() {
    this.props.clearTimeout(this.timeoutID)
  }

  closeAndRun = (func) => (evt) => {
    if (evt) evt.stopPropagation()
    this.closePopup()
    func()
  }

  closePopup = () => {
    const { setTimeout, showMiniBagConfirm } = this.props
    this.timeoutID = setTimeout(() => {
      showMiniBagConfirm(false)
      this.timeoutID = undefined
    }, 200)
  }

  render() {
    const { l } = this.context
    const { isShown, products, quantity, openMiniBag, redirectTo } = this.props
    const hydratedProducts = products.map((product) => ({
      ...product,
      quantity,
      totalPrice: parseFloat(product.unitPrice) * quantity,
    }))
    const className = classnames('MiniBagConfirm', {
      'is-visible': isShown && hydratedProducts.length,
    })

    return (
      <div className={className}>
        <ToolTip>
          <div className="MiniBagConfirm-content">
            <button
              className="MiniBagConfirm-closeIcon Modal-closeIcon"
              onClick={this.closePopup}
            >
              <span className="screen-reader-text">Close confirmation</span>
            </button>
            <ul className="MiniBagConfirm-products">
              {hydratedProducts.map((product) => (
                <li key={product.productId}>
                  <OrderProduct {...product} />
                </li>
              ))}
            </ul>
            <Button
              className="MiniBagConfirm-goToCheckout"
              clickHandler={this.closeAndRun(
                redirectTo.bind(null, '/checkout')
              )} // eslint-disable-line react/jsx-no-bind
            >{l`Checkout now`}</Button>
            <Button
              className="MiniBagConfirm-viewBag Button--secondary"
              clickHandler={this.closeAndRun(openMiniBag)}
            >{l`View bag`}</Button>
          </div>
        </ToolTip>
      </div>
    )
  }
}

const sortProducts = (bundleProducts, products) => {
  return sort(({ productId: productOneId }, { productId: productTwoId }) => {
    return (
      bundleProducts.findIndex(propEq('productId', productOneId)) -
      bundleProducts.findIndex(propEq('productId', productTwoId))
    )
  }, products)
}

export default connect(
  ({
    currentProduct: { bundleProducts = [] } = {},
    shoppingBag: { recentlyAdded },
  }) => ({
    isShown: recentlyAdded.isMiniBagConfirmShown,
    quantity: recentlyAdded.quantity,
    products: sortProducts(bundleProducts, recentlyAdded.products),
  }),
  { showMiniBagConfirm, openMiniBag }
)(MiniBagConfirm)

export { MiniBagConfirm as WrappedMiniBagConfirm }
