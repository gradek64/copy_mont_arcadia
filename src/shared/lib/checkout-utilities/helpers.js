import { sum } from 'ramda'
import { formatForRadix } from './../price'

/**
 * Calculate the shopping bag total
 * FIX: the 'total' calculated we receive from the backend is not always updated
 */
export const fixTotal = (subTotal, shippingCost = 0, discounts = []) => {
  // handle commas in european money
  const discountsConverted = discounts.map((x) => formatForRadix(x))
  return (
    parseFloat(formatForRadix(subTotal)) +
    parseFloat(formatForRadix(shippingCost)) -
    sum(discountsConverted.map(parseFloat))
  )
}
