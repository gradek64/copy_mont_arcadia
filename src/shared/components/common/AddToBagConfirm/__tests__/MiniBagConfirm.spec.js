import toJson from 'enzyme-to-json'

import testComponentHelper from 'test/unit/helpers/test-component'
import { mockStoreCreator } from 'test/unit/helpers/get-redux-mock-store'

import MiniBagConfirm, { WrappedMiniBagConfirm } from '../MiniBagConfirm'

const noop = () => {}

describe('<MiniBagConfirm />', () => {
  const requiredProps = {
    showMiniBagConfirm: noop,
    openMiniBag: noop,
  }
  const products = [
    {
      name: 'TALL Black Leigh Jeans',
      unitPrice: '38.00',
      size: '14',
      inStock: true,
      lowStock: false,
      assets: [],
      productId: 29750936,
    },
    {
      name: 'Chain Belted Jacket',
      unitPrice: '65.00',
      size: '10',
      inStock: true,
      lowStock: false,
      assets: [],
      productId: 29752415,
    },
  ]
  const renderComponent = testComponentHelper(WrappedMiniBagConfirm)

  describe('@decorators', () => {
    describe('connect', () => {
      it('should order products based on the current product‘s `bundleProducts` order', () => {
        const store = mockStoreCreator({
          currentProduct: {
            bundleProducts: [
              {
                productId: 29752415,
              },
              {
                productId: 29750936,
              },
            ],
          },
          shoppingBag: {
            recentlyAdded: {
              products,
            },
          },
        })
        const renderComponent = testComponentHelper(MiniBagConfirm)
        const { wrapper } = renderComponent({ store })
        expect(wrapper.prop('products')).toEqual([
          {
            name: 'Chain Belted Jacket',
            unitPrice: '65.00',
            size: '10',
            inStock: true,
            lowStock: false,
            assets: [],
            productId: 29752415,
          },
          {
            name: 'TALL Black Leigh Jeans',
            unitPrice: '38.00',
            size: '14',
            inStock: true,
            lowStock: false,
            assets: [],
            productId: 29750936,
          },
        ])
      })

      it('should not order the products if there is no current product', () => {
        const store = mockStoreCreator({
          shoppingBag: {
            recentlyAdded: {
              products,
            },
          },
        })
        const renderComponent = testComponentHelper(MiniBagConfirm)
        const { wrapper } = renderComponent({ store })
        expect(wrapper.prop('products')).toEqual(products)
      })
    })
  })

  describe('@renders', () => {
    it('should render in default state', () => {
      const { getTree } = renderComponent(requiredProps)
      expect(getTree()).toMatchSnapshot()
    })

    it('should add `is-visible` class if `isShown` prop is `true` and there are `products`', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        isShown: true,
        products,
      })
      expect(wrapper.find('.MiniBagConfirm').hasClass('is-visible')).toBe(true)
    })

    it('should render `products` correctly', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        products,
        quantity: 2,
      })
      expect(toJson(wrapper.find('.MiniBagConfirm-products'))).toMatchSnapshot()
    })
  })

  describe('@lifecycle', () => {
    describe('on componentDidUpdate', () => {
      it('should call `showMiniBagConfirm` after 6 seconds if `isShown` property changes from `false` to `true`', () => {
        const setTimeoutMock = jest.fn(() => 12345)
        const showMiniBagConfirmMock = jest.fn()
        const { instance } = renderComponent({
          ...requiredProps,
          isShown: true,
          setTimeout: setTimeoutMock,
          showMiniBagConfirm: showMiniBagConfirmMock,
        })
        instance.componentDidUpdate({ isShown: false })
        expect(instance.timeoutID).toBe(12345)
        expect(setTimeoutMock.mock.calls[0][1]).toBe(6000)
        // called the supplied timeout callback
        setTimeoutMock.mock.calls[0][0]()
        expect(showMiniBagConfirmMock).toHaveBeenCalledWith(false)
        expect(instance.timeoutID).toBeUndefined()
      })
    })

    describe('on componentWillUnmount', () => {
      it('should clear timeout', () => {
        const clearTimeoutMock = jest.fn()
        const { instance } = renderComponent({
          ...requiredProps,
          clearTimeout: clearTimeoutMock,
        })
        instance.timeoutID = 12345
        instance.componentWillUnmount()
        expect(clearTimeoutMock).toHaveBeenCalledWith(12345)
      })
    })
  })

  describe('@events', () => {
    describe('on ‘Checkout Now’ button  click', () => {
      it('should stop event prapogation', () => {
        const redirectToMock = jest.fn()
        const stopPropagationMock = jest.fn()
        const { wrapper } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
        })
        wrapper.find('.MiniBagConfirm-goToCheckout').prop('clickHandler')({
          stopPropagation: stopPropagationMock,
        })
        expect(stopPropagationMock).toHaveBeenCalled()
      })

      it('should call `showMiniBagConfirm` after `200` milliseconds (and store the timeout ID)', () => {
        const redirectToMock = jest.fn()
        const setTimeoutMock = jest.fn(() => 12345)
        const showMiniBagConfirmMock = jest.fn()
        const { wrapper, instance } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
          setTimeout: setTimeoutMock,
          showMiniBagConfirm: showMiniBagConfirmMock,
        })
        wrapper.find('.MiniBagConfirm-goToCheckout').prop('clickHandler')()
        expect(instance.timeoutID).toBe(12345)
        expect(setTimeoutMock.mock.calls[0][1]).toBe(200)
        // called the supplied timeout callback
        setTimeoutMock.mock.calls[0][0]()
        expect(showMiniBagConfirmMock).toHaveBeenCalledWith(false)
        expect(instance.timeoutID).toBeUndefined()
      })

      it('should redirect to `/checkout`', () => {
        const redirectToMock = jest.fn()
        const { wrapper } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
        })
        wrapper.find('.MiniBagConfirm-goToCheckout').prop('clickHandler')()
        expect(redirectToMock).toHaveBeenCalledWith('/checkout')
      })
    })

    describe('on ‘View Bag’ button click', () => {
      it('should stop event propagation', () => {
        const redirectToMock = jest.fn()
        const stopPropagationMock = jest.fn()
        const { wrapper } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
        })
        wrapper.find('.MiniBagConfirm-viewBag').prop('clickHandler')({
          stopPropagation: stopPropagationMock,
        })
        expect(stopPropagationMock).toHaveBeenCalled()
      })

      it('should call `showMiniBagConfirm` after `200` milliseconds (and store the timeout ID)', () => {
        const redirectToMock = jest.fn()
        const setTimeoutMock = jest.fn(() => 12345)
        const showMiniBagConfirmMock = jest.fn()
        const { wrapper, instance } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
          setTimeout: setTimeoutMock,
          showMiniBagConfirm: showMiniBagConfirmMock,
        })
        wrapper.find('.MiniBagConfirm-viewBag').prop('clickHandler')()
        expect(instance.timeoutID).toBe(12345)
        expect(setTimeoutMock.mock.calls[0][1]).toBe(200)
        // called the supplied timeout callback
        setTimeoutMock.mock.calls[0][0]()
        expect(showMiniBagConfirmMock).toHaveBeenCalledWith(false)
        expect(instance.timeoutID).toBeUndefined()
      })

      it('should call `openMiniBag`', () => {
        const redirectToMock = jest.fn()
        const openMiniBagMock = jest.fn()
        const { wrapper } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
          openMiniBag: openMiniBagMock,
        })
        wrapper.find('.MiniBagConfirm-viewBag').prop('clickHandler')()
        expect(openMiniBagMock).toHaveBeenCalled()
      })
    })

    describe('on ‘Close’ button click', () => {
      it('should call `showMiniBagConfirm` after `200` milliseconds (and store the timeout ID)', () => {
        const redirectToMock = jest.fn()
        const setTimeoutMock = jest.fn(() => 12345)
        const showMiniBagConfirmMock = jest.fn()
        const { wrapper, instance } = renderComponent({
          ...requiredProps,
          redirectTo: redirectToMock,
          setTimeout: setTimeoutMock,
          showMiniBagConfirm: showMiniBagConfirmMock,
        })
        wrapper.find('.Modal-closeIcon').prop('onClick')()
        expect(instance.timeoutID).toBe(12345)
        expect(setTimeoutMock.mock.calls[0][1]).toBe(200)
        // called the supplied timeout callback
        setTimeoutMock.mock.calls[0][0]()
        expect(showMiniBagConfirmMock).toHaveBeenCalledWith(false)
        expect(instance.timeoutID).toBeUndefined()
      })
    })
  })
})
