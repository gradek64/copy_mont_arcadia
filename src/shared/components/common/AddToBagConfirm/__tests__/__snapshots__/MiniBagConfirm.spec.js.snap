// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MiniBagConfirm /> @renders should render \`products\` correctly 1`] = `
<ul
  className="MiniBagConfirm-products"
>
  <li
    key="29750936"
  >
    <Connect(OrderProduct)
      assets={Array []}
      inStock={true}
      lowStock={false}
      name="TALL Black Leigh Jeans"
      productId={29750936}
      quantity={2}
      size="14"
      totalPrice={76}
      unitPrice="38.00"
    />
  </li>
  <li
    key="29752415"
  >
    <Connect(OrderProduct)
      assets={Array []}
      inStock={true}
      lowStock={false}
      name="Chain Belted Jacket"
      productId={29752415}
      quantity={2}
      size="10"
      totalPrice={130}
      unitPrice="65.00"
    />
  </li>
</ul>
`;

exports[`<MiniBagConfirm /> @renders should render in default state 1`] = `
<div
  className="MiniBagConfirm"
>
  <ToolTip>
    <div
      className="MiniBagConfirm-content"
    >
      <button
        className="MiniBagConfirm-closeIcon Modal-closeIcon"
        onClick={[Function]}
      >
        <span
          className="screen-reader-text"
        >
          Close confirmation
        </span>
      </button>
      <ul
        className="MiniBagConfirm-products"
      />
      <Button
        className="MiniBagConfirm-goToCheckout"
        clickHandler={[Function]}
      >
        Checkout now
      </Button>
      <Button
        className="MiniBagConfirm-viewBag Button--secondary"
        clickHandler={[Function]}
      >
        View bag
      </Button>
    </div>
  </ToolTip>
</div>
`;
